### PR TITLE
chore: add rsources stats for dropped events at processor

### PR DESCRIPTION
**Pull request by Sidddddarth**
# Description

- All events failed during the processor's pipeline (failed, filtered) are treated as dropped with respect to rudder-sources stats.
- For dropped events we are not capturing failed records

## Linear Ticket

[Ensure drop events are marked as failed in source [job-status endpoint]](https://linear.app/rudderstack/issue/PIPE-177/ensure-drop-events-are-marked-as-failed-in-source-[job-status)

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: processor/processor.go [hunk 1]
```diff
@@ -2216,7 +2216,8 @@ func (proc *Handle) transformSrcDest(
 			var successCountMetadataMap map[string]MetricMetadata
 			eventsToTransform, successMetrics, successCountMap, successCountMetadataMap = proc.getDestTransformerEvents(response, commonMetaData, eventsByMessageID, destination, transformer.UserTransformerStage, trackingPlanEnabled, transformationEnabled)
 			failedJobs, failedMetrics, failedCountMap := proc.getFailedEventJobs(response, commonMetaData, eventsByMessageID, transformer.UserTransformerStage, transformationEnabled, trackingPlanEnabled)
-			proc.saveFailedJobs(failedJobs)
+			droppedJobs := proc.getDroppedJobs(response, eventList)
+			proc.saveFailedJobs(append(failedJobs, droppedJobs...))
```
Comment (atzoum) [inline processor/processor.go:2220]: I believe that for rudder sources' use cases both failed and dropped jobs in processor's pipeline could be considered as dropped (i.e. non-retryable)

1. We could rename 
  - `proc.saveFailedJobs` to `proc.saveDroppedJobs `
  - `rsources.NewFailedJobsCollector` to `rsources.NewDroppedJobsCollector`
  - `rsources.JobsFailed` to `rsources.JobsDropped`
3. Dropped jobs failed keys shouldn't be captured
4. It would be better to report rsources dropped jobs during `Store`, once for all scenarios as the last operation of the method, after storing router, batchrouter & proc_errors

File: processor/events_response.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/processor/transformer"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/samber/lo"
+)
+
+func (proc *Handle) getDroppedJobs(response transformer.Response, eventsToTransform []transformer.TransformerEvent) []*jobsdb.JobT {
+	// each messageID is one event when sending to the transformer
+	inputMessageIDs := lo.Map(eventsToTransform, func(e transformer.TransformerEvent, _ int) string {
+		return e.Metadata.MessageID
```
Comment (atzoum) [inline processor/events_response.go:15]: I am confused why processor's pipeline operates on the assumption that messageIDs are unique. Dedup is a feature that can be turned off, right? Am I missing some other important fact with respect to message ids and the processor's pipeline?

File: services/debugger/transformation/transformationStatusUploader.go [hunk 3]
```diff
@@ -390,24 +391,6 @@ func (h *Handle) processRecordTransformationStatus(tStatus *TransformationStatus
 					IsError:          true,
 				})
 			}
-		} else if failedEvent.Metadata.MessageID != "" {
```
Comment (atzoum) [inline services/debugger/transformation/transformationStatusUploader.go:393]: how is this change relevant to the scope of the pull request?

File: processor/processor.go [hunk 4]
```diff
@@ -2225,7 +2236,7 @@ func (proc *Handle) transformSrcDest(
 			var successCountMetadataMap map[string]MetricMetadata
 			eventsToTransform, successMetrics, successCountMap, successCountMetadataMap = proc.getDestTransformerEvents(response, commonMetaData, eventsByMessageID, destination, transformer.UserTransformerStage, trackingPlanEnabled, transformationEnabled)
 			failedJobs, failedMetrics, failedCountMap := proc.getFailedEventJobs(response, commonMetaData, eventsByMessageID, transformer.UserTransformerStage, transformationEnabled, trackingPlanEnabled)
-			proc.saveFailedJobs(failedJobs)
```
Comment (atzoum) [inline processor/processor.go:2228]: Are we intentionally excluding `failedJobs` from rsources stats? Should we treat them as dropped jobs instead?

Comment (Sidddddarth) [reply processor/processor.go:2228]: oh yeah
missed it among other changes.

File: services/debugger/transformation/transformationStatusUploader.go (hunk 3 above)
Comment (Sidddddarth) [reply services/debugger/transformation/transformationStatusUploader.go:393]: not really, but it's the same pattern that we saw in processor, so removed it.

**Review by atzoum — APPROVED** [review]

File: processor/events_response.go (hunk 2 above)
Comment (cisse21) [reply processor/events_response.go:16]: Yeah that is correct but dedup ideally should never be turned off

Also for many of the downstream destinations like Warehouse they do operate under the assumption that the primary key is the messageID which is unique

**Review by cisse21 — APPROVED** [review]
